docs/polkadot/networking/3-attestation.md

```diff
@@ -1,54 +1,268 @@
+
+The protocol runs in several phases and stages. Every node acts both in the distributor and distributee role, but not every role is active in every stage. A summary follows:
+
+Phase | Distributors | Distributees
+----- | ------------ | ------------
+P1SA  | Y            | N
+P1SB  | Y            | Y
+P2SA  | Y            | N
+P2SB  | N            | Y
+
+Note that there is also background activity, as described below.
+
+## Protocol phase 1: initial distribution
+
+Every validator is both a distributor of C pieces and a distributee (recipient) of C pieces [TODO: correction, explanation]. Every piece has one source parachain and one main target storer, and so we can index pieces with a tuple (c_s, (c_t, i_t)).
+
+In phase 1, pieces are distributed by the source parachain validators (c_s, \*) to the main target storers. This happens in two stages. Stage A is where most of the material is distributed, and stage B acts as a backup mechanism for anything that was missed during stage A.
+
+**Stage A**
+
+As a distributor, each validator (c, i) attempts to send the relevant pieces for their parachain's block (c, (c', i)), to everyone else on their ring (c', i) for all c' != c. Conversely as a distributee, each validator expects to receive their relevant pieces (c', (c, i)) for other parachains' blocks, from everyone else on their ring (c', i) for all c' != c.
+
+In more detail:
+
+Each distributor (c, i) will, with parallelism = C / 4, for s in [0..C), try to send the relevant piece to target t = (c+s, i) [TBX S1 #2]. C / 4 comes from our estimate that T_b ~= 4 * T_L.
+
+Trials are done with a timeout, slightly larger than T_l. Sending is via QUIC. In order for it to be treated as a success, it should include an acknowledgement of receipt. Note this is orthogonal from the gossiped receipts which include a validator signature; by contrast this transport-level receipt can be assumed to be already protected by QUIC transport authentication.
+
+If a gossiped receipt is received at any point during the whole process, for a target for a piece, then we can interpret that to mean that the target obtained the piece from a different sender in the meantime, and we should cancel the sending attempt with success.
+
+**Stage B**
+
+As a distributee, if after a grace period we still haven't received our piece (c', (c, i)) from a parachain validator (c', i), then we will ask the other validators (c', i') for all i' != i in that parachain for the piece (c', (c, i)), load-balanced as described in more detail below.
+
+This gives the distributee a more direct level of control over obtaining their own pieces.
+
+As a distributor, if after our own stage A process is finished, we have received fewer than ceil(N/3) of the receipts of another ring i' - i.e. receipts for piece (c, (c', i')) from peer (c', i') for all c' - then we will begin the stage A process for the other ring i' too, load-balanced as described in more detail below.
```

Comment:
I am confused. The *receipt* here is a *Stage A* receipt or a gossip receipt? In the former case why would a validator expect to receive receipts of validators from a different ring?

Reply:
It's a gossip receipt. I'm not sure what you mean by a "stage A" receipt, these gossip receipts are the same whether sent in stage A or B.

There is a transport-level ACK to confirm a direct-send was received correctly but that's a separate idea and only mentioned in the lower-level context of a QUIC connection.

Reply:
Understood. How about the following chance to make it more clear?

```suggestion
As a distributor, if after our own stage A process is finished, we have received fewer than ceil(N/3) of the receipts via the gossip overlay of another ring i' - i.e. receipts for piece (c, (c', i')) from peer (c', i') for all c' - then we will begin the stage A process for the other ring i' too, load-balanced as described in more detail below.
```

Reply:
I don't think this is necessary - "receipt" and "gossiped receipt" are used interchangeably many times already in this document, and I don't see much point fixing all 20+ usages of it one way or the other.